core/src/main/scala/io/chrisdavenport/keypool/KeyPool.scala

```
@@ -231,20 +232,26 @@ object KeyPool{
       val idleCount_ = idleCount - toDestroy.length
       (PoolMap.open(idleCount_, toKeep), toDestroy)
     }
+
+    val sleep = Timer[F].sleep(5.seconds).void
+
     // Wait 5 Seconds
     def loop: F[Unit] = for {
-      _ <- Timer[F].sleep(5.seconds)
+      now <- Timer[F].clock.monotonic(NANOSECONDS)
       _ <- {
-        kpVar.modify{
-          case p@PoolClosed() => (p, Applicative[F].pure(()))
+        kpVar.tryModify {
+          case p@PoolClosed() => (p, Applicative[F].unit)
           case p@PoolOpen(idleCount, m) =>
-            if (m.isEmpty) (p, loop) // Not worth it to introduce deadlock concerns when hot loop is 5 seconds
+            if (m.isEmpty) (p, Applicative[F].unit) // Not worth it to introduce deadlock concerns when hot loop is 5 seconds
             else {
-              val (m_, toDestroy) = findStale(idleCount,m)
-              (m_, toDestroy.traverse_(r => destroy(r._1, r._2).attempt.void))
+              val (m_, toDestroy) = findStale(now, idleCount,m)
+              (m_, toDestroy.traverse_(r => destroy(r._1, r._2))
```

Dropped the `attempt`, because `.attempt.void` swallows any errors.  The `keepRunning` will give us another bite at the apple in five seconds.

---

No, because we removed them from the reference here. So if 1 fails, we don't want the other destroy actions to fail.